src/main/java/io/stargate/sgv2/jsonapi/api/model/command/clause/update/RenameOperation.java

```diff
@@ -0,0 +1,85 @@
+package io.stargate.sgv2.jsonapi.api.model.command.clause.update;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import io.stargate.sgv2.jsonapi.exception.ErrorCode;
+import io.stargate.sgv2.jsonapi.exception.JsonApiException;
+import io.stargate.sgv2.jsonapi.util.PathMatch;
+import io.stargate.sgv2.jsonapi.util.PathMatchLocator;
+import java.util.ArrayList;
+import java.util.List;
+
+/** Implementation of {@code $rename} update operation used to rename fields of documents. */
+public class RenameOperation extends UpdateOperation<RenameOperation.Action> {
+  private RenameOperation(List<Action> actions) {
+    super(actions);
+  }
+
+  public static RenameOperation construct(ObjectNode args) {
+    List<Action> actions = new ArrayList<>();
+    var it = args.fields();
+    while (it.hasNext()) {
+      var entry = it.next();
+      String srcPath = validateUpdatePath(UpdateOperator.RENAME, entry.getKey());
+      JsonNode value = entry.getValue();
+      if (!value.isTextual()) {
+        throw new JsonApiException(
+            ErrorCode.UNSUPPORTED_UPDATE_OPERATION_PARAM,
+            ErrorCode.UNSUPPORTED_UPDATE_OPERATION_PARAM.getMessage()
+                + ": $rename requires STRING parameter for 'to', got: "
+                + value.getNodeType());
+      }
+      String dstPath = validateUpdatePath(UpdateOperator.RENAME, value.textValue());
+      actions.add(new Action(PathMatchLocator.forPath(srcPath), PathMatchLocator.forPath(dstPath)));
+    }
+    return new RenameOperation(actions);
+  }
+
+  @Override
+  public boolean updateDocument(ObjectNode doc) {
+    boolean modified = false;
+    for (Action action : actions) {
+      PathMatch src = action.sourceLocator().findIfExists(doc);
+      JsonNode value = src.removeValue();
+      if (value != null) {
+        // Mongo does not allow "renaming" of Array values
+        if (src.contextNode().isArray()) {
+          throw new JsonApiException(
+              ErrorCode.UNSUPPORTED_UPDATE_OPERATION_PATH,
+              ErrorCode.UNSUPPORTED_UPDATE_OPERATION_PATH.getMessage()
+                  + ": $rename does not allow ARRAY field as source ('"
+                  + action.sourceLocator()
+                  + "')");
+        }
+
+        // If there is a value will be a modification (since source value will
+        // disappear), regardless of whether target changes
+        modified = true;
+        PathMatch dst = action.targetLocator().findOrCreate(doc);
```

Comment:
There is document which has address sub-document like {"address" : {"address line" : "123 main st", "city" : "NYC", "state": "NY"}} 
If the update request is to rename city to town, the sub-document fields (city and state) order will change as {"address" : {"address line" : "123 main st", "state": "NY", "city" : "NYC"}}? Is this the expected behavior?

Reply:
@maheshrajamani As far as I can see, mongo docs state (and Atlas seem to implement it like so) that `$rename` is implemented as "$unset followed by $set"; and $set appends at the end.

So I think it would indeed be expected that field "moves" the way you describe; `city` removed, `town` appended.